traffic_ops/client_tests/cachegroups.go

```diff
@@ -0,0 +1,99 @@
+package main
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+import "net/http"
+
+import "github.com/apache/trafficcontrol/lib/go-tc"
+import "github.com/apache/trafficcontrol/traffic_ops/traffic_ops_golang/api"
+
+// Static Cache Group fields
+var EDGE_CACHEGROUP_ID = 1
+var EDGE_CACHEGROUP = "Edge"
+var EDGE_CACHEGROUP_SHORT_NAME = "Edge"
+var EDGE_CACHEGROUP_LATITUDE = 0.0
+var EDGE_CACHEGROUP_LONGITUDE = 0.0
+var EDGE_CACHEGROUP_PARENT_NAME = "Mid" // NOTE: This places a hard requirement on the `cachegroups` implementation - must have a `MID_LOC` Cache Group named "Mid"
+var EDGE_CACHEGROUP_PARENT_ID = 2       // NOTE: This places a hard requirement on the `cachegroups` implementation - must have a `MID_LOC` Cache Group identified by `2`
+var EDGE_CACHEGROUP_FALLBACK_TO_CLOSEST = true
+var EDGE_CACHEGROUP_LOCALIZATION_METHODS = []tc.LocalizationMethod{
+	tc.LocalizationMethodCZ,
+	tc.LocalizationMethodDeepCZ,
+	tc.LocalizationMethodGeo,
+}
+
+var MID_CACHEGROUP_ID = 2
+var MID_CACHEGROUP = "Mid"
+var MID_CACHEGROUP_SHORT_NAME = "Mid"
+var MID_CACHEGROUP_LATITUDE = 0.0
+var MID_CACHEGROUP_LONGITUDE = 0.0
+var MID_CACHEGROUP_FALLBACK_TO_CLOSEST = true
+var MID_CACHEGROUP_LOCALIZATION_METHODS = []tc.LocalizationMethod{
+	tc.LocalizationMethodCZ,
+	tc.LocalizationMethodDeepCZ,
+	tc.LocalizationMethodGeo,
+}
+
+var CACHEGROUPS = []tc.CacheGroupNullable{
```

Comment:
These would be the static test cases, but the issue here would be that they would have to be kept up to date with any new/deleted fields in the structs. I was kind of envisioning dynamic test cases too that would automatically stay up to date by using reflection to enumerate all of the current fields and populate them with randomly chosen (or statically-chosen value for strings, ints, floats, etc) values to emit from the API.

For example, I'm envisioning iterating over the struct fields and populating them based on what type the field is. We could just say "if it's a string type, use 'foo'. If it's a pointer to a string type, use a pointer to the string 'foo'. If it's an integer type, use 42." Or, we could say "if it's a string type, randomly generate a 8-char string to populate the field with. Or, generate a random float value for float fields". I'm not sure how difficult that would be or if it's even possible with Golang's reflection capabilities, but I think it would be really valuable to keep this Mock TO up to date automatically.

Reply:
Some of that behavior can lead to invalid or self-inconsistent fields, like `Type: "asdfg",` and `Latitude: 400.6`. We could make constraints by field, but at some point that becomes more complex than defining static defaults - and with essentially the same effect. Plus, making it dynamic/random could harm testability. At the least, I want this thing to be able to be used for tests of ORT, and that means it's beneficial to have static configfile output. But if profile/server/cdn names/ids and associated parameter names/ids/values/locations are random/dynamic then they could conflict with the static output of the configfiles. I don't think that's a problem today, but setting up a self-inconsistent testing environment seems like a bad idea to me. In order to make that deterministic, we'd need to build configfile generation logic into the mock server to be maintained separately from TO's configfile generation logic, and probably do the same for any tests using the mock server.

It's unfortunate because I share your concern about maintaining structure twice, but this seems like the lesser of the two evils, and less work to maintain.

Reply:
I like the suggestion. The more we can decouple the need to modify the mock TO, the easier (and more likely) it will be kept up to date. 

To keep the responses consistent across instances/versions of the mock server, instead of randomly generating perhaps we could derive the value of the field from properties of that field like name/type.

Reply:
> _"To keep the responses consistent across instances/versions of the mock server, instead of randomly generating perhaps we could derive the value of the field from properties of that field like name/type."_

The problem with that is it's going to have to look something like:
```go
if field.Name == "Latitude" {
    field.Value = 0.0
}
```
and I'm just not sure how that's meaningfully different than 
```go
Latitude: 0.0
```
Besides that the former is less likely to cause compilation errors. But imo, it'd be better to get a compilation error, so you immediately know what's wrong, because either way you need to change something if the field name changes from `Latitude` to `Foo`.

Reply:
What if it was instead (with equally weak Go-fu reflection)

  for field: range fields {
    field.Name: CreateValue(field.Name, field.Type)
  }

Reply:
That's a good point, I was mostly just handwaving away invalid values based upon the types, but I suppose we do have clients that would expect to get a valid value to parse into specific types (thinking about the Golang client here). Clients like the Python TO client and TP don't really parse the values into specific types that are different than the json types, so it doesn't really matter what value they get back (as long as the "primitive" types are the same).

I guess, what would the value in Mock TO be if it was just all 100% statically returned values? At that point, we don't really need to use the go-tc structs; we can just create static JSON files that are served up by a simple http server in a simple directory structure:
```
api/
    1.2/
        cachegroups <---static json file
        asns <--- static json file
```
where `curl http://localhost:30003/api/1.2/cachegroups` would return that static json. It would essentially be accomplishing the same thing, except for the go struct's marshalling/unmarshalling code (which I think is already well-tested elsewhere).

If Mock TO isn't integrated enough with Real TO to not require maintenance every time real TO changes something, I think it will lose its value quickly as it gets more and more out of date.

Reply:
Right, its an impossible line to walk- trying to build a server that both understands the context of a specific field, but does not require updates when new fields are added. 

I'm still fuzzy on the purpose of this code. Have you considered writing a UT suite against ORT and then mocking out the specific calls it will make on a function by function basis? When new functionality gets added to ORT, a new test gets added as well.

Reply:
> _"we can just create static JSON files that are served up by a simple http server in a simple directory structure"_

Yeah we could. My concern with that is double-maintenance of values. Like, right now, I have the parent of the edge cache group set to a pointer to the name of the mid cache group. That way you only have to change it once. If they're in JSON, we lose the ability to make those connections, unless we're going to do text replacements which once again more or less ends with us having all the values statically defined in the Go code anyway.

> _"Have you considered writing a UT suite against ORT and then mocking out the specific calls it will make on a function by function basis?"_
That's more or less what I'm doing. Thing is, we have ORT.pl and my baby, ORT.py. I want the tests to be independent of implementation, i.e. "after a SYNCDS run of an ORT implementation pointed at the Mock TO, the exact file system and process table changes should be this:". Building tests into the Perl code limits my ability to to test Python, and vice-versa. Basically the entire reason I wanted a mock TO in the first place was to figure out what gaps there still are between the two implementations.

What I've done here is build a fake HTTP service that ORT requires to run, which is the first step in writing the actual tests. By separating them, I can make a more general-purpose testing tool that could be used by more general-purpose clients, like the Apache-TrafficControl Python package and the Golang client @rawlinp mentioned.

Reply:
> If they're in JSON, we lose the ability to make those connections, unless we're going to do text replacements which once again more or less ends with us having all the values statically defined in the Go code anyway.

We'd only need to make the connection once when creating the static files, and I'm not sure why the connections would ever need to be updated after that. An upside of just using static json files instead would be that you could just query a running TO instance, find/replace stuff you think should be redacted, and the static files would be ready to serve to mock the entire API (instead of writing a bunch of code). If you wanted to test specific changes to the static responses, all you'd have to do is manually update the static files instead of changing hard-coded values in the server, recompiling, and restarting the server.

Reply:
I'm not super worried about restarting the server, since this should never be any part of production.

As far as updates, you _probably_ won't need to update them _after an exhaustive set_ is determined. Like we probably want a ton more cache servers so that our tests cover a bunch of mutually-exclusive cases. Which means updating cache groups and parents and profiles and parameters etc. etc. Then also, whenever something changes like "oh this thing that used to allow arbitrary text now has a restricted format" you would need to grep/sed through all the JSON. Which may or may not be a big deal. I think it is, but I could be convinced otherwise.

Reply:
> I'm not super worried about restarting the server, since this should never be any part of production.

That's not what I'm worried about either. I was just pointing out that it's easier to just modify a file than it is to modify some source code, recompile, and restart a server for your changes to be realized.

With Mock TO I was mostly interested in the idea of it being able to stay up to date automatically with Real TO by reusing the same Go structs and other traffic_ops_golang code, but if it's all going to be static then I don't see a really compelling reason why it has to be Go when it can just be static json files and a simple http server.

Reply:
The only reasons I really have are because it will allow us to to always provide the same "alerts" objects as a real instance - without ever needing to update -, and re-use of values that makes adding new endpoints easier and is hard to replicate with JSON. I don't really feel strongly about JSON vs Go structs (because secretly I hate all of those global declarations), so if you feel the benefits of using JSON outweigh the benefits of the static Go structs I'll change it.

Reply:
If there will likely be repetition between the endpoints (for example if you have 3 DS'), then staying with the structs will reduce duplication. You could declare a base struct then copy it and modify as needed. When we add more DS fields, only the base needs to change. In contrast, if we had 3 separate DS JSON files, each one of those would need an update- tedious.

Reply:
I'm gonna put this on the mailing list to see if we can reach some kind of consensus.

Reply:
There's a reason why I haven't explicitly -1'd this PR and have just been commenting on it. I was hoping to get some features out of it that I had originally proposed, but that doesn't necessarily have to hold up the statically-defined implementation. I do wish we had a representative dataset like Kabletown 2.0 that we could leverage for the static-Mock TO use case, but maybe we'll get there eventually.

Reply:
Kabletown 2.0 would be great - but for it to be useful the API responses need to be deterministic IMO. Until things like The Duplicate Origin Problem<sup>TM</sup> are solved, a static dataset isn't enough because you can't predict what the result of a request will be.

Reply:
Since we'd control the static Kabletown 2.0 dataset, we could simply keep the origins unique, which would mean that the API responses would be deterministic.